ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp

```diff
@@ -38,6 +33,44 @@ namespace ngraph
     {
         namespace reference
         {
+            namespace details
+            {
+                template <typename T>
+                void dot(const T* arg0,
+                         const T* arg1,
+                         T* out,
+                         const Shape& arg0_shape,
+                         const Shape& arg1_shape,
+                         const Shape& out_shape)
+                {
+                    std::fill(out, out + shape_size(out_shape), static_cast<T>(0));
+                    size_t arg0_rank = arg0_shape.size();
+                    size_t arg1_rank = arg1_shape.size();
+
+                    // 2D inputs shapes are interpreted as {I, K} x {K, J}
+                    // If first input is 1D tensor of shape {K}, it is interpreted as {1, K}
+                    // If second input is 1D tensor of shape {K}, it is interpreted as {K, 1}
+                    size_t I_dim = arg0_rank == 1 ? 1 : arg0_shape[arg0_rank - 2];
+                    size_t J_dim = arg1_rank == 1 ? 1 : arg1_shape[arg1_rank - 1];
+                    size_t K_dim =
+                        arg1_rank == 1 ? arg1_shape[arg1_rank - 1] : arg1_shape[arg1_rank - 2];
+
+                    size_t a_idx = 0, b_idx = 0, out_idx = 0;
+                    for (size_t i = 0; i < I_dim; ++i)
+                    {
+                        for (size_t k = 0; k < K_dim; ++k)
+                        {
+                            a_idx = i * K_dim + k;
+                            for (size_t j = 0; j < J_dim; ++j)
+                            {
+                                b_idx = k * J_dim + j;
+                                out_idx = i * J_dim + j;
+                                out[out_idx] += arg0[a_idx] * arg1[b_idx];
+                            }
+                        }
+                    }
```

Comment:
In idiomatic C++ it is preferred to declare variables in the nearest place of its first usage. 

```
for (size_t i = 0; i < I_dim; ++i) { 
  for (size_t k = 0; k < K_dim; ++k) {
        const size_t a_idx = i * K_dim + k; 
        for (size_t j = 0; j < J_dim; ++j) { 
            const size_t b_idx = k * J_dim + j; 
            const size_t out_idx = i * J_dim + j; 
            out[out_idx] += arg0[a_idx] * arg1[b_idx]; 
        }
    }
}
```

Reply:
I heard that declaration before nested loops is kind of optimization, but I've moved it as you suggest.

Reply:
Oh, got your point. I've checked with godbolt and generated assembly looks similar (I've counted lines of code) so I wouldn't be concerned about speed here :) Code: https://godbolt.org/z/315aae 

But in any case, the best way would be to measure first and only then make such an optimizations. We have open discussion with @ilyachur how to introduce benchmarks for operations (including reference), so maybe in the near future we'll have an easy way to measure such things.